Improve typing for unresolved managers

ljodal opened the pull request:
This changes the logic when encountering an unresolvable manager class. Instead of adding it as a `Manager` we create a subclass of `Manager` that has `fallback_to_any=True` set. Similarly a `QuerySet` class is created that also has fallbacks to `Any`. This allows calling custom methods on the manager and querysets without getting type errors.

ljodal committed:
Improve typing for unresolved managers

This changes the logic when encountering an unresolvable manager class.
Instead of adding it as a `Manager` we create a subclass of `Manager`
that has `fallback_to_any=True` set. Similarly a `QuerySet` class is
created that also has fallbacks to `Any`. This allows calling custom
methods on the manager and querysets without getting type errors.

flaeppe commented on mypy_django_plugin/transformers/models.py:
```diff
@@ -83,6 +85,81 @@ def get_generated_manager_info(self, manager_fullname: str, base_manager_fullnam
+
+        # Check if we've already created a fallback queryset class for this
+        # module, and if so reuse that.
+        queryset_info = self.lookup_typeinfo(f"{self.model_classdef.info.module_name}.{name}")
+        if queryset_info and queryset_info.metadata.get("django", {}).get("any_fallback_queryset"):
+            return queryset_info
```
While _highly_ unlikely, not saying you'll have to fix this, there could be name collisions and `add_new_class_for_current_module` guarantees unique names. So `add_new_class_for_current_module` _has a risk_ of inserting `{self.model_classdef.info.module_name}.{name}1`

But I suppose that'd just lead to always creating and not reusing info.

There's also a risk of finding and using the colliding symbol, if it happens to be a `TypeInfo`, but yeah.

---

Same goes for the manager fallback creation

ljodal replied:
I'm aware of the chance for a collision, but the only effect should be that a new manager is created every time rather than being reused. The metadata should ensure that we don't accidentally use a manager/queryset that's _not_ created here?

flaeppe replied:
Yeah, true. The metadata will ensure we don't use something else, didn't think of that.

ljodal commented on mypy_django_plugin/transformers/models.py:
```diff
@@ -83,6 +85,81 @@ def get_generated_manager_info(self, manager_fullname: str, base_manager_fullnam
+        if not related_manager:
+            django_metadata["from_queryset_managers"] = {
+                f"django.db.models.manager.{manager_info.fullname}": manager_info.fullname,
+            }
```
I _think_ this bit can be removed, as we're not going to try to do a dynamic lookup of this manager?

flaeppe replied:
This only happens on final iteration, right? And the `UnknownManager` won't ever match a runtime manager either.

I'm thinking the case is like:

Runtime manager type is something like: `manager.from.third.party.ThirdPartyManager`, and we'll insert an `{current_module.fullname}.UnknownManager` in place of it. In general, if the case is that a runtime manager won't ever have the possibility to match what we generate there's no need to store any metadata, I think.

---

Side note, if you decide to keep it; `f"django.db.models.manager.{manager_info.name}"` should be the metadata value

ljodal replied:
Yes, it's only on the final iteration, and I'm gonna remove this. Wasn't entirely clear to me what this was used for when I wrote it, but it's clear now that this is not needed :)

flaeppe commented on tests/typecheck/fields/test_related.yml:
```diff
@@ -680,7 +680,7 @@
                 # Note, that we cannot resolve dynamic calls for custom managers:
                 class Transaction(models.Model):
-                    objects = BaseManager.from_queryset(TransactionQuerySet)
+                    objects = BaseManager.from_queryset(TransactionQuerySet)()
```
IMO this test is missing a `reveal_type(Transaction.objects)`. I don't think `reveal_type(Transaction.objects.custom())` by itself tells enough of the story(?)

flaeppe reviewed (commented):
Could we also add a test case having two different unresolvable managers, in the same module, that'll also reveal their model attribute type?

ljodal committed:
Fix manager created and improve a test

ljodal committed:
Fix row type of FallbackQuerySet

Because this inherits from _QuerySet, not QuerySet, it needs to have two
parameters

sobolevn reviewed (approved):
Thank you!
